src/EventStore.Core/Services/Transport/Enumerators/Enumerator.AllSubscriptionFiltered.cs

```diff
@@ -71,462 +59,380 @@ public class AllSubscriptionFiltered : IAsyncEnumerator<ReadResponse> {
 
-						default:
-							Fail(ReadResponseException.UnknownError.Create(completed.Result));
-							return;
+					if (liveEvent.ResolvedEvent.HasValue) {
+						// this is the next event to send towards the client. send it and update `checkpoint`
+						await SendEventToSubscription(liveEvent.ResolvedEvent.Value, ct);
+						checkpoint = liveEvent.ResolvedEvent.Value.OriginalPosition!.Value;
+					} else if (liveEvent.Checkpoint.HasValue) {
+						// this is a valid checkpoint that we can send towards the client. send it and update `checkpoint`
+						await SendCheckpointToSubscription(liveEvent.Checkpoint.Value, ct);
+						checkpoint = new TFPos(liveEvent.Checkpoint.Value.CommitPosition, liveEvent.Checkpoint.Value.PreparePosition);
 					}
 				}
+
+				throw new Exception($"Unexpected error: live events channel for subscription {_subscriptionId} to $all:{_eventFilter} completed without exception");
 			}
 
-			private void GoLive(Position startPosition) {
-				var liveEvents = Channel.CreateBounded<Message>(BoundedChannelOptions);
-				var caughtUpSource = new TaskCompletionSource<Position>();
-				var liveMessagesCancelled = 0;
+			private Task<TFPos> CatchUp(TFPos checkpoint, CancellationToken ct) {
+				Log.Verbose(
+					"Subscription {subscriptionId} to $all:{eventFilter} is catching up from checkpoint {position}",
+					_subscriptionId, _eventFilter, checkpoint);
 
-				Log.Information(
-					"Live subscription {subscriptionId} to $all running from {position}...",
-					_subscriptionId, startPosition);
+				var checkpointIntervalCounter = 0L;
+				var catchupCompletionTcs = new TaskCompletionSource<TFPos>();
 
-				_bus.Publish(new ClientMessage.FilteredSubscribeToStream(Guid.NewGuid(), _subscriptionId,
-					new ContinuationEnvelope(OnSubscriptionMessage, _semaphore, _cancellationToken), _subscriptionId,
-					string.Empty, _resolveLinks, _user, _eventFilter, (int)_checkpointInterval, (int)_checkpointIntervalCounter));
+				// this is a safe use of AsyncTaskEnvelope. Only one call to OnMessage will be running
+				// at any given time because we only expect one reply and that reply kicks off the next read.
+				AsyncTaskEnvelope envelope = null;
+				envelope = new AsyncTaskEnvelope(OnMessage, ct);
 
-				Task.Factory.StartNew(PumpLiveMessages, _cancellationToken);
+				ReadPage(checkpoint, envelope, ct);
 
-				async Task PumpLiveMessages() {
-					await caughtUpSource.Task;
+				return catchupCompletionTcs.Task;
 
-					await _channel.Writer.WriteAsync(new ReadResponse.SubscriptionCaughtUp(), _cancellationToken);
+				async Task OnMessage(Message message, CancellationToken ct) {
+					try {
+						if (message is ClientMessage.NotHandled notHandled &&
+						    TryHandleNotHandled(notHandled, out var ex))
+							throw ex;
+
+						if (message is not ClientMessage.FilteredReadAllEventsForwardCompleted completed)
+							throw ReadResponseException.UnknownMessage
+								.Create<ClientMessage.ReadAllEventsForwardCompleted>(message);
```

Comment:
```suggestion
								.Create<ClientMessage.FilteredReadAllEventsForwardCompleted>(message);
```
probably just an oversight

Reply:
thanks, yup i missed that